analyzer/consensus/genesis.go

```diff
@@ -0,0 +1,553 @@
+// Conversion of the genesis state into SQL statements.
+
+package consensus
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/iancoleman/strcase"
+	"github.com/oasisprotocol/oasis-core/go/common/entity"
+	"github.com/oasisprotocol/oasis-core/go/common/node"
+	genesis "github.com/oasisprotocol/oasis-core/go/genesis/api"
+	registry "github.com/oasisprotocol/oasis-core/go/registry/api"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+
+	"github.com/oasisprotocol/oasis-indexer/log"
+)
+
+const bulkInsertBatchSize = 1000
+
+// MigrationGenerator generates migrations for the Oasis Indexer
+// target storage.
+type MigrationGenerator struct {
+	logger *log.Logger
+}
+
+// NewMigrationGenerator creates a new migration generator.
+func NewMigrationGenerator(logger *log.Logger) *MigrationGenerator {
+	return &MigrationGenerator{logger}
+}
+
+// WriteGenesisDocumentMigrationOasis3 creates a new migration that re-initializes all
+// height-dependent state as per the provided genesis document.
+func (mg *MigrationGenerator) WriteGenesisDocumentMigrationOasis3(w io.Writer, document *genesis.Document) error {
```

Comment:
This will only be `oasis3`?

Reply:
We'll see. Theoretically, oasis4 could be entirely different. In practice, I expect we'll be able to slightly refactor this function and reuse most of it for oasis4 too. For now, it is true that it only ever parses oasis3.